esp8266-fastled-webserver/Fields.cpp

```diff
@@ -17,479 +17,526 @@
-  hueIncMax = value.toInt();
-  return value;
-}
+    String getSHueBpm() {
+      return String(sHueBpm);
+    }
+    String setSHueBpm(String value) {
+      sHueBpm = value.toInt();
+      return value;
+    }
 
-String getSHueBpm() {
-  return String(sHueBpm);
-}
-String setSHueBpm(String value) {
-  sHueBpm = value.toInt();
-  return value;
-}
+    String getSHueMin() {
+      return String(sHueMin);
+    }
+    String setSHueMin(String value) {
+      sHueMin = value.toInt();
+      return value;
+    }
 
-String getSHueMin() {
-  return String(sHueMin);
-}
-String setSHueMin(String value) {
-  sHueMin = value.toInt();
-  return value;
-}
+    String getSHueMax() {
+      return String(sHueMax);
+    }
+    String setSHueMax(String value) {
+      sHueMax = value.toInt();
+      return value;
+    }
 
-String getSHueMax() {
-  return String(sHueMax);
-}
-String setSHueMax(String value) {
-  sHueMax = value.toInt();
-  return value;
-}
+  }
 
+  inline namespace Options {
+    void getPatterns(JsonVariant dst) {
+      for (uint8_t i = 0; i < patternCount; i++) {
+        dst.add(patterns[i].name);
+      }
+    }
 
-typedef String (*FieldSetter)(String);
-typedef String (*FieldGetter)();
-struct Field {
-  String name;
-  String label;
-  String type;
-  uint8_t min;
-  uint8_t max;
-  FieldGetter getValue;
-  FieldGetter getOptions;
-  FieldSetter setValue;
-};
-
-// passing array reference works fine, but need to make the function a template
-// to capture the array size... on the positive side, no need to pass `count` parameter
-template <size_t N>
-Field  getField(String name, const Field (&fields)[N]) {
-  for (uint8_t i = 0; i < N; i++) {
-    Field field = fields[i];
-    if (field.name == name) {
-      return field;
+    void getPalettes(JsonVariant dst) {
+      for (uint8_t i = 0; i < paletteCount; i++) {
+        dst.add(paletteNames[i]);
+      }
     }
   }
-  return Field();
-}
-template <size_t N>
-String getFieldValue(String name, const Field (&fields)[N]) {
-  Field field = getField(name, fields);
-  if (field.getValue) {
-    return field.getValue();
-  }
-  return String();
-}
-template <size_t N>
-String setFieldValue(String name, String value, const Field (&fields)[N]) {
-  Field field = getField(name, fields);
-  if (field.setValue) {
-    return field.setValue(value);
-  }
-  return String();
-}
-template <size_t N>
-String getFieldsJson(const Field (&fields)[N]) {
-  String json = "[";
-
-  for (uint8_t i = 0; i < N; i++) {
-    Field field = fields[i];
 
-    json += "{\"name\":\"" + field.name + "\",\"label\":\"" + field.label + "\",\"type\":\"" + field.type + "\"";
+  typedef String (*FieldSetter)(String);
+  typedef String (*FieldGetter)();
+  typedef void (*FieldOptions)(JsonVariant);
+  struct Field {
+    const String name;
+    const String label;
+    Field_t type;
+    uint8_t min;
+    uint8_t max;
+    FieldGetter  getValue;
+    FieldOptions getOptions;
+    FieldSetter  setValue;
+  };
```

Comment:
I do believe that for memory and performance reasons, one could change all these strings (except name and label) to uint32_t (FieldGetter / Fieldsetter, maybe min and max...).

The only place where the fieldvalues are required as String is a webserver response. This could either work with the actual value directly or converison to String in the callback function.

The label uses a "string getter" (first entry in Fields) but there is certainly a workaround possible for that one as it is the only label.
The "legacy" solidColor beign returned as comma-separated rgb, the 32bit (24 bit) representation can be used which would require a small adoption the the web app as well.

Reply:
You are **_absolutely_** correct.  See prior responses ... explicit choice not to change JSON format with this PR.